### DISCUSSION TITLE
Setting up TypeScript & webpack-dev-server

### DISCUSSION THREAD

**Discussion by gyenabubakar** [webpack/webpack#14929] [Q&A]
Use this repo to reproduce the issue I'm facing: https://github.com/gyenabubakar/setting-up-ts

Currently, this is what I get:

![image](https://user-images.githubusercontent.com/26672526/145209244-f8ca069c-83ad-4f57-a39d-44e3c2b8b9b9.png)

Nothing is working at all.

I want `webpack-dev-server` to serve with the `index.html` in the root directory. The HTML file references a `/dist/bundle.js` file which should be in memory because of `webpack-dev-server`.

I thought that `output.publicPath` in `webpack.config.js` would do the trick but it doesn't.

**Comment by alexander-akait** [comment]
Do you use `html-webpack-plugin`?

**Reply by gyenabubakar** [reply]
I just did and added it to webpack.config, but still nothing.

In the messages logged by `webpack-dev-server` I realised it was serving files from `__dirname + '/public'` so I moved the HTML file there and the page was served.

But it still can't load the `bundle.js`.

**Reply by alexander-akait** [reply]
Give me time, I will look at repo

**Reply by alexander-akait** [reply]
If you don't want to use `HtmlWebpackPlugin` please use and serve `index.html` from root of project:
```
devServer: {
        static: './'
}
```

But I don't recommend to store public files in root of project, my recommendation is move index.html file to `public` directory and using:
```
devServer: {
  static: './public'
}
```

If you need to use `HtmlWebpackPlugin`, please use `index.html` as template for `HtmlWebpackPlugin`, remove `<script type="module" src="/dist/bundle.js" defer></script>` from your HTML, `HtmlWebpackPlugin` by default inject chunks, note - you don't need `defer`, because all script with `type="module"` are deferred, also if you need module output please use https://webpack.js.org/configuration/output/#outputmodule.

Also you need to set valid `publicPath`, in your case just remove `publicPath` or set it to `publicPath: "auto"`, you will have relative URLs in your bundle in this case. Also remove `contentBase`, it is the old option for webpack-dev-server v3, it was renamed to the `static` option, by default dev server serve `public` directory. 

Shorty I recommend:

Setup to use `public`/`dist` directory (as you prefer):
- set `output.path` to `path.resolve(__dirname, 'public')` (or `dist`, by default used webpack)
- remove `output.publicPath`, so you will have relative URLs
- do not override the `static` option (if you use `dist`, you will need to setup `static: './dist'` for dev server)
- use `index.html` in the `template` option for `HtmlWebpackPlugin` ( and remove `script` tag), i.e.

```
plugins: [
  new HtmlWebpackPlugin({
    template: "index.html"
  })
}
```

If you need `favicon.ico` and other files, you can use `copy-webpack-plugin`.

Feel free to feedback
